appsec/appsec.go

```diff
@@ -0,0 +1,196 @@
+	eventChan          chan *appsectypes.SecurityEvent
+	wg                 sync.WaitGroup
+	cfg                *Config
+	instrumentationIDs []dyngo.EventListenerID
+}
+
+type Logger interface {
+	Warn(string, ...interface{})
+	Error(string, ...interface{})
+}
+
+func NewAgent(client *http.Client, logger Logger, cfg *Config) (*Agent, error) {
+	intakeClient, err := intake.NewClient(client, cfg.AgentURL)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Hostname == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			logger.Warn("unable to look up hostname: %v", err)
+		} else {
+			cfg.Hostname = hostname
+		}
+	}
+
+	if cfg.Service.Name == "" {
+		name, err := os.Executable()
+		if err != nil {
+			logger.Warn("unable to look up the executable name: %v", err)
+		} else {
+			cfg.Service.Name = filepath.Base(name)
+		}
+	}
+
+	if cfg.MaxBatchLen <= 0 {
+		cfg.MaxBatchLen = defaultMaxBatchLen
+	}
+
+	if cfg.MaxBatchStaleTime <= 0 {
+		cfg.MaxBatchStaleTime = defaultMaxBatchStaleTime
+	}
+
+	return &Agent{
+		eventChan: make(chan *appsectypes.SecurityEvent, 1000),
+		client:    intakeClient,
+		cfg:       cfg,
+	}, nil
+}
+
+func (a *Agent) Start(ctx context.Context) {
+	a.run(ctx)
+}
+
+func (a *Agent) Stop(gracefully bool) {
+	dyngo.Unregister(a.instrumentationIDs)
+	// Stop the batching goroutine
+	close(a.eventChan)
+	// Possibly wait for the goroutine to gracefully stop
+	if !gracefully {
+		return
+	}
+	a.wg.Wait()
+}
+
+func (a *Agent) run(ctx context.Context) {
+	a.instrumentationIDs = append(a.instrumentationIDs, httpprotection.Register()...)
+	a.instrumentationIDs = append(a.instrumentationIDs, a.listenSecurityEvents()...)
+
+	a.wg.Add(1)
+	go func() {
+		defer a.wg.Done()
+
+		globalEventCtx := []appsectypes.SecurityEventContext{
+			appsectypes.ServiceContext{
+				Name:        a.cfg.Service.Name,
+				Version:     a.cfg.Service.Version,
+				Environment: a.cfg.Service.Environment,
+			},
+			appsectypes.TagContext(a.cfg.Tags),
+			appsectypes.TracerContext{
+				Runtime:        "go",
+				RuntimeVersion: runtime.Version(),
+				Version:        version,
+			},
+			appsectypes.HostContext{
+				Hostname: a.cfg.Hostname,
+				OS:       fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+			},
+		}
+
+		eventBatchingLoop(ctx, a.client, a.eventChan, globalEventCtx, a.cfg)
+	}()
+}
+
+type IntakeClient interface {
+	SendBatch(context.Context, api.EventBatch) error
+}
+
+func eventBatchingLoop(ctx context.Context, client IntakeClient, eventChan <-chan *appsectypes.SecurityEvent, globalEventCtx []appsectypes.SecurityEventContext, cfg *Config) {
+	batch := make([]*appsectypes.SecurityEvent, 0, cfg.MaxBatchLen)
+	timer := time.NewTimer(time.Hour)
+	timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Return immediately as the context is now done so the http client cannot be used anymore
+			return
+
+		case event, ok := <-eventChan:
+			if event != nil {
+				batch = append(batch, event)
+			}
+			if !ok {
+				timer.Stop()
```

Comment:
The timer is running if the batch has at least one element. So it actually should move to the next `if len(batch) > 0` condition block 👍

Note that I've written as many tests of the batching loop as I could think of in `batching_test.go`.

Reply:
Changed in 5258dd5db1db52bbc98772befac5d1893d037076 where I simplified a lot the code of the loop, and I stop the timer all the time now, even when it was stopped for the sake of code simplicity instead of adding an extra value to know if the timer is stopped.

Reply:
Let me know if anything in the new loop code.